### ISSUE TITLE
Fakes that wrap another object are not considered to be equal to themselves

### ISSUE THREAD

**Issue by twilek** [FakeItEasy/FakeItEasy#1630]
I have found a behaviour that I might belive is a bug, here is how to reproduce:		

```
var myClass = A.Fake<MyClass>(o => o.Wrapping(new MyClass()));
myClass.Equals(myClass); // This will not be true
// Which will result in that assertions like this will fail using Fluent Assertions:
myClass.Should().Be(myClass);
```

MyClass can be any class, for example an empty one like this ```public class MyClass{}```

**Comment by thomaslevesque** [comment]
Hi @twilek,

Thanks for reporting this.

This is indeed unfortunate... However it can be easily explained. When you make a wrapping fake, methods that are not explicitly configured call the corresponding method on the wrapped instance.

So, if I rewrite your code like this to give more explicit names:
```csharp
var real = new MyClass();
var fake = A.Fake<MyClass>(o => o.Wrapping(real));
```
When you call `fake.Equals(fake)`, it actually calls `real.Equals(fake)`, which is false... On the other hand, `fake.Equals(real)` calls `real.Equals(real)`, which is true.

This code
```csharp
Console.WriteLine("fake.Equals(fake): " + fake.Equals(fake));
Console.WriteLine("fake.Equals(real): " + fake.Equals(real));
Console.WriteLine("real.Equals(fake): " + real.Equals(fake));
Console.WriteLine("real.Equals(real): " + real.Equals(real));
```
Outputs
```
fake.Equals(fake): False
fake.Equals(real): True
real.Equals(fake): False
real.Equals(real): True
```

Anyway, I agree that this behavior isn't ideal, and even though there's a logical explanation, I think we should treat it as a bug. Maybe we should special-case `Equals`:
- `fake.Equals(fake)` should return true
- `fake.Equals(real)` should return false, since `real.Equals(fake)` returns false.

@blairconrad what do you think?

**Comment by thomaslevesque** [comment]
In terms of implementation, I think we just need to skip the `WrappedObjectRule` for the `Equals` method.

**Comment by twilek** [comment]
Ok, thank you for the detailed explanation. For now I have solved it in my tests with adding an .Unwrap() method like this `result.Should().Be(expected.Unwrap())` where Unwrap is an extension method that looks like this:

```
public static object Unwrap(this object obj)
{
	return Castle.DynamicProxy.ProxyUtil.GetUnproxiedInstance(obj);
}
```

**Comment by thomaslevesque** [comment]
That doesn't seem to work for me... `GetUnproxiedInstance` just returns the same instance.

Anyway, a clean workaround would be something like this:

```csharp
A.CallTo(() => fake.Equals(A<object>._)).ReturnsLazily((object o) => ReferenceEquals(o, fake));
```

**Comment by blairconrad** [comment]
Hi, @twilek. I understand your difficulties, but I'm not yet settled on a course of action. (Nor entirely whether it's "a bug".)

To address @thomaslevesque's idea of skipping the `WrappedObjectRule` for `Equals`, I assume that means we'd have regular Fake semantics for equality, which is essentially reference semantics (technically it's "do I have the same `FakeManager`?").

But what if the wrapped object's `Equals` method was defined? Maybe it's equal to `3`. Then we wouldn't be preserving that behaviour. I think it's important to preserve the behaviour that was explicitly asked for in the Fake configuration. I think we'd be better keeping the current configuration than using default Fake semantics for `Equals` (and `GetHashCode`, because we'd need to align their behaviour).

I was initially thinking that we could replace the behaviour with `ReferenceEquals || wrapped.Equals`, essentially. It makes the assumption that `wrapped.Equals` returns `true` when the compared-to object is `this`. I don't think that's unreasonable. But would this make the behaviour too complicated to explain to the user?

Note that we'd almost certainly end up with an `Equals` method where
1.  `fake.Equals(wrappedThing) == true`
2. `fake.Equals(fake) == true`
3.  `wrappedThing.Equals(wrappedthing) == true`
4.  `wrappedThing.Equals(fake) == false` (unless `wrappedThing.Equals` is overridden and some some sort of value-based semantics)

But currently  we probably have 1, 3, and 4 and !2. Unless `wrappedThing.Equals` is overridden and approaches value-based semantics, in which case `fake` might already equal `fake`.

And I don't think we can do _anything_ about 4, so we'd be breaking the `Equals` method's symmetry (instead of breaking breaking reflexivity).

**Comment by blairconrad** [comment]
I think @thomaslevesque's workaround of actually defining `Equals`'s behaviour is good, but you might need to define `GetHashCode`'s as well.

Of course, if you're only worried about comparisons _in the test_, you might be able to get away with `actual.Should().BeSameAs(expected)`.

**Comment by twilek** [comment]
Ah, yes, `actual.Should().BeSameAs(expected)` actually works for me and is a much nicer way of doing it than my ugly workaround, thank you for that!

**Comment by thomaslevesque** [comment]
In fact, my workaround can be simplified:

```csharp
A.CallTo(() => fake.Equals(fake)).Returns(true);
```

> But what if the wrapped object's `Equals` method was overridden? Maybe it's equal to `3`. Then we wouldn't be preserving that behaviour. I think it's important to preserve the behaviour that was explicitly asked for in the Fake configuration, and that it would be better to keep the current configuration than to use default Fake semantics for `Equals` (and `GetHashCode`, because they must agree).

That's a good point.

> I was initially thinking that we could replace the behaviour with `ReferenceEquals || wrappedThing.Equals`. It makes the assumption that `wrappedThing.Equals` returns `true` when the compared-to object is `this`. I don't think that's unreasonable. But would this make the behaviour too complicated to explain to the user?

It would make `fake.Equals(wrappedThing)` true in all cases, which I don't think make sense, unless equality has been overriden by `wrappedThing`. Maybe we could check if `Equals` was overriden? A bit hacky, but it might work.

> Note that we'd almost certainly end up with an `Equals` method where
> 
>     1. `fake.Equals(wrappedThing) == true`
>     2. `fake.Equals(fake) == true`
>     3. `wrappedThing.Equals(wrappedthing) == true`
>     4. `wrappedThing.Equals(fake) == false` (unless `wrappedThing.Equals` is overridden and has some some sort of value-based semantics)

It's annoying to introduce an asymmetry... But I don't see a way around it.

> I think @thomaslevesque's workaround of actually defining `Equals`'s behaviour is good, but you might need to define `GetHashCode`'s as well.

In fact, it doesn't matter. The only constraint is that objects have the same hashcode if they're equal. The workaround makes `fake` equal to itself, which satisfies this constraint, since obviously `fake.GetHashCode() == fake.GetHashCode()`.

**Comment by thomaslevesque** [comment]
> It's annoying to introduce an asymmetry... But I don't see a way around it.

Or maybe there is...
- If the argument is `fake`, return true
- If the argument is `wrappedThing`, return `wrappedThing.Equals(fake)`

This way, both `fake.Equals(wrappedThing)` and `wrappedThing.Equals(fake)` will return:
-  false if `Equals` is not overridden (reference equality)
- true if `Equals` is overridden to implement value semantics

**Comment by thomaslevesque** [comment]
Ah, in fact the solution I described in my last message only works if the value semantics is based on virtual members... Otherwise the wrapping fake doesn't reflect the properties of the wrapped object.

Which leads me to another point: the main (only?) reason to override `Equals` is to implement equality with value semantics. And I can't think of a realistic scenario where one would need a wrapping fake of an object with value semantics. I'm not saying it doesn't exist, but it's probably pretty rare... So maybe we're just splitting hairs for nothing.

**Comment by blairconrad** [comment]
> Ah, in fact the solution I described in my last message only works if the value semantics is based on virtual members... Otherwise the wrapping fake doesn't reflect the properties of the wrapped object.

Then again, I would hope that in most cases, the faked type is an interface. I am not sure why you'd fake a class just to wrap an instance of the class. Again, there's probably a use case, but I expect it's rare.

Actually, now I'm curious. @twilek, in your example, you're wrapping a class with a faked class instead of just faking the class with the [CallsBaseOptions](https://fakeiteasy.readthedocs.io/en/stable/calling-base-methods/) option.
Is that just to simplify the example?

Back to @thomaslevesque's comment, there's would also be complications if the Fake is explicitly configured. Then the wrapped object's properties also might not agree with the wrapping fake.

Still, I think it's most important to give a good experience for unconfigured fakes and even configured fakes that are more "normal" or "sane", based on our arbitrary and capricious whims. :wink:

I think  [your proposal](https://github.com/FakeItEasy/FakeItEasy/issues/1630#issuecomment-535478383) is a good one. There will be cases where combinations of virtual or non-virtual methods and an overridden `Equals` method could cause some surprises, but we'll get the proper reflexive behaviour and proper behaviour for an default `Equals`.  I'm in.

**Comment by twilek** [comment]
@blairconrad Yes, it is indeed a simplified example. I'm writing a combat engine to a game. The engine is based around a state machine, and I use the same setup for several different tests and in one test I check if the current combatant (a state object) is the one that has been targeted (`result.().Should().Be(state.CurrentTurnOrder.CurrentCombatant);`) and in several other test I want the CurrentCombatant to behave like the original object, not the fake so first I didn't even fake it, but then I added a test case where I wanted to fake a behavior in CurrentCombatant and then I changed the object into a wrapped object, and then after doing that the first test failed. So that was the reason I found this behavior.

**Comment by thomaslevesque** [comment]
>  so first I didn't even fake it, but then I added a test case where I wanted to fake a behavior in CurrentCombatant and then I changed the object into a wrapped fake, and then after doing that the first test started to fail.

In this scenario I think you'd be better off using a "normal" fake with `CallsBaseOptions`. It would avoid all the issues discussed in this thread. And anyway, as mentioned by @blairconrad earlier, what you need here is reference equality (`actual.Should().BeSameAs(expected)`).

**Comment by twilek** [comment]
Of reasons maybe a bit too long to explain here the CurrentCombatant can not be a fake though in most of my tests, but in short the object is a state holder that also holds other objects with states (combatant stats, combatant abilities etc) and what these tests actually do is to test changes in the state tree when things happening so I want them to be concrete objects rather than fakes. But then I added one test where it actually would be easier to test what I wanted if I just faked a behavior in CurrentCombatant instead of using the concrete implementation so that was the reason I just changed it into a wrapped fake. But sure, I could actually do a fake just for this test instead of using the concrete object created in the setup, but I thought it was just easier to make the already existing one to a wrapped object :) And yes, BeSameAs(expected) is indeed the more correct way in either way (fake or not). Thanks!

**Comment by thomaslevesque** [comment]
> I thought it was just easier to make the already existing one to a wrapped object :)

In your case, I think it wouldn't change anything. By default, the fake would behave the same as the concrete class, unless you explicitly configure its behavior.

**Comment by twilek** [comment]
The base methods are dependent on several other concrete objects set from the constructor though, so the base methods won't function as expected if I just fake the object if I not also rewire these dependencies to the expected objects. And that is of course possible, and how I usually use to deal with situations like this, but now I already had a concrete object created so I decided to just use a wrapped fake instead and ended up finding this behavior (I've in fact been using fakeiteasy since it was initially released for almost ten years or so and this is one of the few times ever I have been using wrapped fakes, so believe me when I say that this is not a problem I usually have with fakeiteasy :)).

**Comment by blairconrad** [comment]
@twilek, we released 5.3.0-beta.1 with this fix. Are you interested in trying it out before we release 5.3.0?

**Comment by twilek** [comment]
I just tried it out. Seems to work well, now I get the expected result both with .BeSameAs(expected) as well as .Be(expected). All my other tests also passes. Great work!

**Comment by blairconrad** [comment]
Great! Thanks for confirming, @twilek.

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 5.3.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/5.3.0).

Thanks, @twilek. Look for your name in the release notes! 🥇
